### PR TITLE
[ESDB-106-15] Always replicate the chunk header

### PR DESCRIPTION
**Pull request by shaan1337**
Changed: Always replicate the chunk header

This PR causes followers to always replicate the chunk header from the leader. It is a pre-requisite for the encryption at rest plugin. Prior to this PR, the chunk header was replicated only for scavenged chunks.

This change has an additional benefit: it's easy for sysadmins to compare files on a leader and a follower by using standard tools like `md5sum`, `sha1sum`, etc. without needing to exclude the first 128 bytes from the comparison (as previously, the chunk IDs were different on the leader and followers)

### CODE REVIEW

**Comment by shaan1337** [comment]
Please don't merge yet, as some additional tests are probably needed.

File: src/EventStore.Core/Services/Replication/LeaderReplicationService.cs
Comment (shaan1337) [inline src/EventStore.Core/Services/Replication/LeaderReplicationService.cs:404]: To verify: is this always `false` ? as in some cases, this block may be entered when the chunk is scavenged

Comment (shaan1337) [reply src/EventStore.Core/Services/Replication/LeaderReplicationService.cs:404]: fixed in: https://github.com/EventStore/EventStore/pull/4265

this PR needs to be rebased once the above PR is merged.

**Review by hayley-jean — CHANGES_REQUESTED** [review]
The replicated chunks seem to lose the version in their filenames when they are replicated.
e.g. `chunk-000001.000001` on the leader is replicated as `chunk-000001.000000` on the follower.

File: src/EventStore.Core/Cluster/ReplicationPartials.cs
Comment (shaan1337) [reply src/EventStore.Core/Cluster/ReplicationPartials.cs]: @hayley-jean nice catch! i've just pushed a fix for it, the previous code wasn't populating the old chunk header field at all for scavenged chunks, thereby causing this exception (i haven't tested it yet to confirm if it's fixed though)

**Comment by shaan1337** [comment]
> The replicated chunks seem to lose the version in their filenames when they are replicated. e.g. `chunk-000001.000001` on the leader is replicated as `chunk-000001.000000` on the follower.

nice find! I believe that this bug already existed prior to this PR due to this call in SwitchChunk:
https://github.com/EventStore/EventStore/blob/aa6c0a2381fc590a8cd76ca7942d21b72a86363c/src/EventStore.Core/TransactionLog/Chunks/TFChunkManager.cs#L191

it could easily be fixed by specifying a minimum value of 1 for the chunk version (although the versions wouldn't necessarily match with the leader)

it's quite important to fix it as some places in the code assume that a chunk with version 0 hasn't been scavenged:
e.g on startup, it assumes that the next chunk number is just incremented by one:
https://github.com/EventStore/EventStore/blob/aa6c0a2381fc590a8cd76ca7942d21b72a86363c/src/EventStore.Core/TransactionLog/Chunks/TFChunkEnumerator.cs#L64

**Comment by shaan1337** [comment]
> > The replicated chunks seem to lose the version in their filenames when they are replicated. e.g. `chunk-000001.000001` on the leader is replicated as `chunk-000001.000000` on the follower.
> 
> nice find! I believe that this bug already existed prior to this PR due to this call in SwitchChunk:
> 
> https://github.com/EventStore/EventStore/blob/aa6c0a2381fc590a8cd76ca7942d21b72a86363c/src/EventStore.Core/TransactionLog/Chunks/TFChunkManager.cs#L191
> 
> it could easily be fixed by specifying a minimum value of 1 for the chunk version (although the versions wouldn't necessarily match with the leader)
> 
> it's quite important to fix it as some places in the code assume that a chunk with version 0 hasn't been scavenged: e.g on startup, it assumes that the next chunk number is just incremented by one:
> 
> https://github.com/EventStore/EventStore/blob/aa6c0a2381fc590a8cd76ca7942d21b72a86363c/src/EventStore.Core/TransactionLog/Chunks/TFChunkEnumerator.cs#L64

sorry, my bad, it is actually a bug introduced in this PR, i'll push a fix for it

File: src/EventStore.Core/Cluster/ReplicationPartials.cs
Comment (shaan1337) [reply src/EventStore.Core/Cluster/ReplicationPartials.cs]: confirming that the fix works

**Review by timothycoleman — APPROVED** [review]
Nice looks good.
I've tested backwards and forwards compatibility for replication

**Pull request merged**
